datafusion/src/optimizer/remove_duplicate_filters.rs

```diff
@@ -0,0 +1,611 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Remove duplicate filters optimizer rule
+
+use crate::execution::context::ExecutionProps;
+use crate::logical_plan::LogicalPlan;
+use crate::logical_plan::{lit, Expr};
+use crate::optimizer::optimizer::OptimizerRule;
+use crate::optimizer::utils;
+use crate::optimizer::utils::optimize_explain;
+use crate::scalar::ScalarValue;
+use crate::{error::Result, logical_plan::Operator};
+
+/// Remove duplicate filters optimizer.
+/// # Introduction
+/// It uses boolean algebra laws to simplify or reduce the number of terms in expressions.
+///
+/// Filter: #b Gt Int32(2) And #b Gt Int32(2)
+/// is optimized to
+/// Filter: #b Gt Int32(2)
+pub struct RemoveDuplicateFilters {}
+
+fn expr_contains<'a>(expr: &'a Expr, needle: &'a Expr) -> bool {
+    match expr {
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } => expr_contains(left, needle) || expr_contains(right, needle),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } => expr_contains(left, needle) || expr_contains(right, needle),
+        _ => expr == needle,
+    }
+}
+
+fn as_binary_expr<'a>(expr: &'a Expr) -> Option<&'a Expr> {
+    match expr {
+        Expr::BinaryExpr { .. } => Some(expr),
+        _ => None,
+    }
+}
+
+fn operator_is_boolean(op: &Operator) -> bool {
+    op == &Operator::And || op == &Operator::Or
+}
+
+fn is_one<'a>(s: &'a Expr) -> bool {
+    match s {
```

Comment:
These are really good use cases for the `matches!`  macro https://doc.rust-lang.org/std/macro.matches.html

Reply:
Ok, nice macro (I did not know it).

Sorry if i'm being too clumsy, but how can I use it if I have to guard two different cases (each one with a different variable/type) ?
```
matches!(s,
   Scalar::Int8(1) |
   Scalar::Int16(1) |
   Scalar::Float32(Some(v)) if *v == 1. |
   Scalar::Float64(Some(v)) if *v == 1. )
```

Seems that maches! only allows one last guard, and the existing binding must have the same type in all alternatives.

Is there any other way ?

Reply:
A `matches!` returns a boolean so you can also use `||` / boolean or to cover the float cases and keep the rest in a single `matches!`

Reply:
Or it might be better to just use `|` patterns (without `matches!`) in the code.

Reply:
I like that last option :)